lib/pbench/cli/server/shell.py

```diff
@@ -24,12 +27,23 @@ def main():
     except (ConfigFileNotSpecified, BadConfig) as e:
         print(e)
         sys.exit(1)
+    logger = get_pbench_logger(__name__, server_config)
     try:
         host = str(server_config.get("pbench-server", "bind_host"))
         port = str(server_config.get("pbench-server", "bind_port"))
+        db = str(server_config.get("Postgres", "db_uri"))
         workers = str(server_config.get("pbench-server", "workers"))
-    except (NoOptionError, NoSectionError) as e:
-        print(f"{__name__}: ERROR: {e.__traceback__}")
+
+        # Multiple gunicorn workers will attempt to connect to the DB; rather
+        # than attempt to synchronize them, detect a missing DB (from the
+        # postgres URI) and create it here.
+        if not database_exists(db):
+            logger.info("Postgres DB {} doesn't exist", db)
+            create_database(db)
+            logger.info("Created DB {}", db)
+        Database.init_db(server_config, logger)
```

Comment:
The only problem I see in initializing the db here is that our code runs gunicorn with a python subprocess module so it spawns a separate process [line 49-58] for our flask app. Therefore any initialization here may not be visible to the process thats is running our flask app.

Reply:
The point is not the in-memory initialization, but the Postgres initialization. We need to create the named DB if it doesn't exist, and we need to create all the tables if they don't exist. It turns out that SQLAlchemy/Postgres have synchronization issues if multiple clients (such as we spawn through gunicorn) attempt to do this at the same time.

Doing it here, when we're single-threaded, avoids those issues.

Reply:
Indeed; I'll amend the comment.